cpp/src/arrow/type.cc

```diff
@@ -608,6 +611,415 @@ std::string DictionaryType::ToString() const {
 
 std::string NullType::ToString() const { return name(); }
 
+// ----------------------------------------------------------------------
+// FieldRef
+
+FieldRef::FieldRef(Indices indices) : impl_(std::move(indices)) {
+  DCHECK_GT(util::get<Indices>(impl_).size(), 0);
+}
+
+FieldRef::FieldRef(std::vector<FieldRef> children) {
+  // flatten children
+  struct Visitor {
+    void operator()(std::string&& name) { *out++ = FieldRef(std::move(name)); }
+
+    void operator()(Indices&& indices) { *out++ = FieldRef(std::move(indices)); }
+
+    void operator()(std::vector<FieldRef>&& children) {
+      for (auto& child : children) {
+        util::visit(*this, std::move(child.impl_));
+      }
+    }
+
+    std::back_insert_iterator<std::vector<FieldRef>> out;
+  };
+
+  std::vector<FieldRef> out;
+  Visitor visitor{std::back_inserter(out)};
+  visitor(std::move(children));
+
+  DCHECK(!out.empty());
+  DCHECK(std::none_of(out.begin(), out.end(),
+                      [](const FieldRef& ref) { return ref.IsNested(); }));
+
+  if (out.size() == 1) {
+    impl_ = std::move(out[0].impl_);
+  } else {
+    impl_ = std::move(out);
+  }
+}
+
+Result<FieldRef> FieldRef::FromDotPath(const std::string& dot_path_arg) {
+  if (dot_path_arg.empty()) {
+    return Status::Invalid("Dot path was empty");
+  }
+
+  std::vector<FieldRef> children;
+
+  util::string_view dot_path = dot_path_arg;
+
+  auto parse_name = [&] {
+    std::string name;
+    for (;;) {
+      auto segment_end = dot_path.find_first_of("\\[.");
+      if (segment_end == util::string_view::npos) {
+        // dot_path doesn't contain any other special characters; consume all
+        name.append(dot_path.begin(), dot_path.end());
+        dot_path = "";
+        break;
+      }
+
+      if (dot_path[segment_end] != '\\') {
+        // segment_end points to a subscipt for a new FieldRef
+        name.append(dot_path.begin(), segment_end);
+        dot_path = dot_path.substr(segment_end);
+        break;
+      }
+
+      if (dot_path.size() == segment_end + 1) {
+        // dot_path ends with backslash; consume it all
+        name.append(dot_path.begin(), dot_path.end());
+        dot_path = "";
+        break;
+      }
+
+      // append all characters before backslash, then the character which follows it
+      name.append(dot_path.begin(), segment_end);
+      name.push_back(dot_path[segment_end + 1]);
+      dot_path = dot_path.substr(segment_end + 2);
+    }
+    return name;
+  };
+
+  while (!dot_path.empty()) {
+    auto subscipt = dot_path[0];
```

Comment:
Should be "subscript" everywhere (there are other occurrences).

Reply:
Thanks!